### PR TITLE
fix: underflow during datetime->nanos conversion

### LINKED ISSUE #1289: Utc nanoseconds do not always roundtrip
# Reproducer

## Code
```rust
#[test]
fn nano_roundrip() {
    let nanos = i64::MIN + 2;
    let dt = Utc.timestamp_nanos(nanos);
    let nanos2 = dt.timestamp_nanos();
    assert_eq!(nanos, nanos2);
}
```

## Expected Result
Test passes.

## Actual Result
Test panics when calling `dt.timestamp_nanos()` (so we don't even get to the actual comparison:

`value can not be represented in a timestamp with nanosecond precision.`

# Technical Background
`dt.timestamp()` is `-9223372037`, however `nanos / 1_000_000_000` is `-9223372036`, hence I conclude that this IF branch was taken:

https://github.com/chronotope/chrono/blob/b64cedc6a8758d15265702895d917045ce1eebf8/src/offset/mod.rs#L427-L430

Hence this `checked_mul` overflows:

https://github.com/chronotope/chrono/blob/b64cedc6a8758d15265702895d917045ce1eebf8/src/naive/datetime/mod.rs#L482

~My guess is that the conversion routine (first code block) isn't entirely accurate.~ The conversion is alright, I rather think when converting back we have to be more careful.

### PR DESCRIPTION
**Pull request by crepererum**
Fixes #1289.

### CODE REVIEW

**Review by djc — APPROVED** [review]
Thanks for the thorough test case!

File: src/datetime/tests.rs [hunk 1]
```diff
@@ -1486,3 +1486,36 @@ fn locale_decimal_point() {
...
+        BILLION,
+        BILLION + 1,
+        i64::MAX - BILLION - 1,
+        i64::MAX - BILLION,
+        i64::MAX - BILLION + 1,
+        i64::MAX - 2,
+        i64::MAX - 1,
+        i64::MAX,
+    ] {
+        println!("nanos: {nanos}");
```
Comment (pitdicker) [inline src/datetime/tests.rs:1516]: Maybe remove this `println!`? Or can you change it so that the test compiles on Rust 1.57 (our MSRV)?

Comment (djc) [reply src/datetime/tests.rs:1516]: Just needs `println!("nanos: {}", nanos)`.

Comment (crepererum) [reply src/datetime/tests.rs]: Adjusted the println to be compatible w/ Rust 1.57. I think keeping it helps debugging a failing tests due to the many cases[^1].

[^1]: I wish there would be a slightly nicer story for parametrized tests in "standard Rust" though.

Comment (djc) [reply src/datetime/tests.rs]: (Hopefully [there will be](https://epage.github.io/blog/2023/06/iterating-on-test/) in the not-too-distant future.)

**Pull request merged**